torch/distributed/pipelining/PipelineSchedule.py

```diff
@@ -685,33 +682,29 @@ def __init__(
 
         self.n_local_stages = len(stages)
         self.rank = stages[0].group_rank
-
-    def _step_microbatches(
-        self,
-        arg_mbs: Optional[List] = None,
-        kwarg_mbs: Optional[List] = None,
-        target_mbs: Optional[List] = None,
-        losses: Optional[List] = None,
-    ):
-        """
-        Operate on the microbatches for interleaved 1f1b schedule (https://arxiv.org/pdf/2104.04473.pdf).
-
-        Highest rank has a warmup (fwd only) count of [len(stages) - 1] * number of PP ranks
-        and each rank away from highest rank adds 2 warmup steps due to:
-            - one happened before highest rank's warmup started,
-            - one waiting for backward result to trickle down from highest rank
-
-        TODO: Interleaved 1F1B does not support using _sorted_batch_p2p()
-        because it requires recvs and sends from different peers
-        to execute in the same coalesced operation. As a result, this schedule does
-        not support models with skip connections.
-        """
-        arg_mbs, kwarg_mbs = self._check_inputs(arg_mbs, kwarg_mbs, target_mbs, losses)
-
-        # increment warmup_steps by 2 for each hop away
-        warmup_steps = (self.n_local_stages - 1) * self.pp_group_size
-        warmup_steps += 2 * ((self.pp_group_size - 1) - self.rank)
-        warmup_steps = min(warmup_steps, self._n_microbatches * self.n_local_stages)
+        self.group = stages[0].group
+
+        # 1. Create the pipeline_order (all ranks do this calculation)
+        # This will be used to keep track of the current state of the entire pipeline
+        # pipeline_order[rank] = [(computation_type, microbatch_index, stage_index), ...]
+        self.pipeline_order: List[
+            List[Optional[Tuple[_ComputationType, int, int]]]
+        ] = []
+        # ========================================================================
+        for rank in range(self.pp_group_size):
+            rank_ops = self._calculate_single_rank_operations(rank)
+            self.pipeline_order.append(rank_ops)
+
+    def _calculate_single_rank_operations(
+        self, rank
+    ) -> List[Optional[Tuple[_ComputationType, int, int]]]:
+        def get_rank_warmup_steps(rank):
+            # Increment warmup_steps by 2 for each hop away
+            warmup_steps = (self.n_local_stages - 1) * self.pp_group_size
+            warmup_steps += 2 * ((self.pp_group_size - 1) - rank)
```

Comment:
nit: comment on the whole calculation. The current comment seems to apply to the second line of code only.

Reply:
yea, better comment would help.  but dont just tell us literally what the math is doing, try to tell us the reasoning.  and perhaps use more temporary variables with instructive names instead of adding comments if that makes it clearer.

The first line `warmup_steps = ` - what is the logical meaning of this line- the number of warmup steps for the last pp stage?  if so, maybe name it `warmups_last_stage`

also, wouldn't the last stage not have any warmups? or im confused

Reply:
Added comments and renamed variables

Reply:
i see. its actually not clear to me why it makes sense to calcuate num_warmups per rank instead of per stage.  wouldn't each stage have a different number of warmups on any given rank?

the variable names used to be 'stage' based, which actually does seem logical to me

Reply:
Previously we were operating at rank level, each rank does determine which stage it will operate on based on the timestep. The timestep should be at rank level since communication is also done at the rank level. I renamed the "steps" variable to be "ops" since operations is more accurate.